### ISSUE TITLE
threading_imgur has a bug

### ISSUE THREAD

**Issue by zmrenwu** [volker48/python-concurrency#4]
Exception in thread Thread-7:
Traceback (most recent call last):
  File "C:\Users\zmrenwu\AppData\Local\Programs\Python\Python35\lib\threading.py", line 914, in _bootstrap_inner
    self.run()
  File "D:/Users/zmrenwu/PycharmProjects/ConcurrentAndParallel/thread.py", line 25, in run
    download_link(directory, link)
  File "D:\Users\zmrenwu\PycharmProjects\ConcurrentAndParallel\download.py", line 21, in download_link
    f.write(image.read())
  File "C:\Users\zmrenwu\AppData\Local\Programs\Python\Python35\lib\http\client.py", line 461, in read
    s = self._safe_read(self.length)
  File "C:\Users\zmrenwu\AppData\Local\Programs\Python\Python35\lib\http\client.py", line 609, in _safe_read
    raise IncompleteRead(b''.join(s), amt)
http.client.IncompleteRead: IncompleteRead(53280 bytes read, 361937 more expected)

It seems image.read() incured.

**Comment by volker48** [comment]
I'll look into it. I see you are on Windows I don't have a Windows machine to test on so I may be unable to reproduce.

Marcus

**Comment by zmrenwu** [comment]
Thanks! I also test it on a linux device. This exception not always throw, I guess it related to my network speed. Then I meet another problem. When all images downloaded, the main thread can't exit. It blocked on queue.join() forever.

**Comment by volker48** [comment]
Sorry, haven't had a chance to check this yet. My wife just had a baby. I'll try and check it this weekend.

Marcus

**Comment by zmrenwu** [comment]
Congratulations! Best wishes to your wife and baby! It doesn't matter. I'm reading the documentation of threading module. Thank you for you help. I learn a lot from your blog post. :)

**Comment by volker48** [comment]
Sorry, just got a chance to take a look at this. I couldn't reproduce the `IncompleteRead` error you are seeing, but like you said thats probably related to the network you are on. I ran the threading code around 20 times with no issues.

I did make a few updates https://github.com/volker48/python-concurrency/commit/2226b9382d05beaf07f41f13da36fbb0d2112b33 in case one of the threads does raise an exception during the download I put the call to `self.queue.task_done()` inside a finally so the main thread won't block on `queue.join()` forever if one of the threads raises an exception.

```
            try:
                download_link(directory, link)
            finally:
                self.queue.task_done()
```
